### PR TITLE
Improve WARM_IP_TARGET handling

### PR DESCRIPTION
**Pull request by mogren**
*Issue #, if available:* #377, #294, #314, #391, #265 and others.

*Description of changes:*
Merge of all fixes that has gone into the 1.4 branch back into master. Add support for actually freeing unused IPs from an attached ENI. Requires the `ec2:UnassignPrivateIPs` permission.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: ipamd/ipamd.go [hunk 1]
```diff
@@ -208,7 +253,11 @@ func (c *IPAMContext) nodeInit() error {
 
 	log.Debugf("Start node init")
 
-	instanceMaxENIs, _ := c.awsClient.GetENILimit()
+	instanceMaxENIs, err := c.awsClient.GetENILimit()
+	if err != nil {
+		log.Errorf("Failed to get ENI limit: %s")
```
Comment (vsiddharth) [inline ipamd/ipamd.go]: Should we not error out in case the limits cannot be determined

File: ipamd/ipamd.go [hunk 2]
```diff
@@ -256,7 +319,7 @@ func (c *IPAMContext) nodeInit() error {
 		ipamdErrInc("nodeInitK8SGetLocalPodIPsFailed", err)
 		// This can happens when L-IPAMD starts before kubelet.
 		// TODO  need to add node health stats here
-		return nil
+		return errors.Wrap(err, "Failed to get running pods!")
```
Comment (vsiddharth) [inline ipamd/ipamd.go]: nit: s/Failed/failed

File: ipamd/ipamd.go [hunk 3]
```diff
@@ -316,12 +379,11 @@ func (c *IPAMContext) getLocalPodsWithRetry() ([]*k8sapi.K8SPodInfo, error) {
 	}
 
 	if err != nil {
-		return nil, err
+		return nil, errors.Wrap(err, "no pods because apiserver not running")
```
Comment (vsiddharth) [inline ipamd/ipamd.go]: nit: add `.` in the end

File: ipamd/ipamd.go (hunk 1 above)
Comment (mogren) [reply ipamd/ipamd.go]: I agree we should handle unknown instance types in a better way, but this change was just to keep it consistent with line 576 were we do the same thing.

**Pull request merged**